From b611466d8ff6ba2e60da83c96387e1c40381ed96 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 3 Jan 2025 18:10:30 +0100
Subject: [PATCH 1429/1695] pta: stm32mp: bsec: implement a BSEC PTA command to
 query the debug state

Implement a BSEC PTA command to query the debug state. This allows
software components that do not have access to the BSEC to get this
information.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 85217aea49d9edc96a169cbe10fdb06a33ad588d pta: stm32mp: bsec: implement a BSEC PTA command to query the debug state]
Change-Id: I67fb0dd67a2c7173dbfee6ec26605c57e3fbe148
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/426946
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/pta/stm32mp/bsec_pta.c            | 10 ++++++++++
 lib/libutee/include/pta_stm32mp_bsec.h |  9 +++++++++
 2 files changed, 19 insertions(+)

diff --git a/core/pta/stm32mp/bsec_pta.c b/core/pta/stm32mp/bsec_pta.c
index 1dc92a729..0a3b88e27 100644
--- a/core/pta/stm32mp/bsec_pta.c
+++ b/core/pta/stm32mp/bsec_pta.c
@@ -275,6 +275,14 @@ static TEE_Result bsec_pta_state(uint32_t pt, TEE_Param params[TEE_NUM_PARAMS])
 	return TEE_SUCCESS;
 }
 
+static TEE_Result bsec_pta_debug_state(void)
+{
+	if (stm32_bsec_self_hosted_debug_is_enabled())
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
 static TEE_Result bsec_pta_invoke_command(void *pSessionContext __unused,
 					  uint32_t cmd_id,
 					  uint32_t param_types,
@@ -290,6 +298,8 @@ static TEE_Result bsec_pta_invoke_command(void *pSessionContext __unused,
 		return bsec_write_mem(param_types, params);
 	case PTA_BSEC_CMD_GET_STATE:
 		return bsec_pta_state(param_types, params);
+	case PTA_BSEC_CMD_GET_DEBUG_STATE:
+		return bsec_pta_debug_state();
 	default:
 		break;
 	}
diff --git a/lib/libutee/include/pta_stm32mp_bsec.h b/lib/libutee/include/pta_stm32mp_bsec.h
index 6fbf9c9f3..ba5322939 100644
--- a/lib/libutee/include/pta_stm32mp_bsec.h
+++ b/lib/libutee/include/pta_stm32mp_bsec.h
@@ -51,6 +51,15 @@
  */
 #define PTA_BSEC_CMD_GET_STATE		0x3
 
+/**
+ * Get the BSEC debug state
+ *
+ * Return codes:
+ * TEE_SUCCESS - BSEC debug enabled
+ * TEE_ERROR_ACCESS_DENIED - BSEC debug disabled
+ */
+#define PTA_BSEC_CMD_GET_DEBUG_STATE	0x4
+
 enum stm32_bsec_pta_sec_state {
 	PTA_BSEC_STATE_SEC_OPEN = 0,
 	PTA_BSEC_STATE_SEC_CLOSE = 1,
-- 
2.47.2

